### ISSUE TITLE
hexadecimal integer gets converted to floating point decimal

### ISSUE THREAD

**Issue by flip111** [sstur/js2php#9]
When i input https://raw.githubusercontent.com/mathiasbynens/utf8.js/master/utf8.js into https://sstur.github.io/js2php/demo/ 

```javascript
if (value >= 0xD800 && value <= 0xDBFF
```

turns into
```php
if ($value >= 55296.0 && $value <= 56319.0
```

Better would be to use the same hexadecimal notation php provides with `0x`

**Comment by sstur** [comment]
Wow, good catch! I didn't think anyone is using this library anymore, it's been quite a few years now.

I'm not sure that I'll find the time to work on a fix for this, but I'm open to a PR.

**Comment by flip111** [comment]
I understand you don't work on it anymore. I was mainly posting this issue as bookkeeping. Maybe somewhere in the future it will be interesting to look into it.

**Comment by flip111** [comment]
What about a flag that let's the user choose? In one mode everything is casted to float. Possibly with another flag to not have casts for literals but add `.0`.

In the other mode js2php keeps track of types in the sense of their notation. So `1` would be type `int` even thought it is really still a float in js. Then when you do `1 + 2` you know the notation type of `1` and `2` was `int`, so you know the result can also be `int`. But as soon as you mix in a float somewhere then result is now "forever a float" since it can not be reliably represented as int. In theory this second mode should be accurate and the first mode can serve as a fall back (just in case it turns out not the be accurate). N.B. one reason why it could not numerically be accurate has to do with the size of the numbers that can be represented.

Then for the boolean comparison operator (strict vs non-strict) you can also keep track of which "notation type" is being compared. Then the conversion could look like this

| JS | PHP |
| --- | --- |
| int === int | int === int |
| float === float | float === float |
| int === float | int == float |
| float === int | float == int |
| anything else === anything else | anything else === anything else |
| anything else == anything else | anything else == anything else |

Note: again, where it says "int" in the JS column this is inferred from the style of notation in JS, not the actual data type!

**Comment by imme64** [comment]
If we would use a flag, this would not only affect the generators but would also be handled in the runtime, increasing complexity a lot. Also all tests would need to be run for all kinds of configurations. And for the user, it is not a real "feature" which he can turn on or off because it only affects internals.

The size of the numbers which can be represented was a very important hint. I played around with PHP, and if a number cannot be represented as integer anymore, it is automatically interpreted as float. `999999999999999999 === 999999999999999999.0` (with 18 digits) is false because int !== float but `9999999999999999999 === 9999999999999999999.0` (19 digits) is true.

Therefore we don't need special handling for large numeric literals, PHP already takes care of it.

I would not try to make the generator very smart regarding detecting numeric types for the boolean comparison operator because in most cases we compare against variables or method results and don't know the type in any case. I would handle it the same way as it is done with _plus(...) method and make type checks during runtime. This would not be the best possible regarding performance but would keep complexity lower because errors can then happen only in runtime and must be searched for in generator and runtime at the same time.

**Comment by sstur** [comment]
This is a super in-depth discussion, you're really diving into the edge cases that I remember going through when I was actively hacking on this back 8 years ago, it brings back fond memories, I dig it!

You've stumbled on some of the reasons I chose to go with the approach of using floats for everything, despite the slightly clumsy PHP output it creates, particularly the difficulty around knowing types at compile-time. Similarly, this is why `_plus()` exists, like you mentioned above. It allows us to check the types at runtime and branch accordingly. In doing so, we pay a runtime cost in addition to the uglier (and larger) generated code. This is unfortunately the cost of accurately replicating JS behavior in PHP without having strong compile-time knowledge of types.

More broadly, I'd refer back to the original objective of this project which is to generate correct, human readable PHP code (with emphasis on correctness over readability). Note that the generated PHP is meant to be readable (ish) but not necessarily _writable_. Meaning it's _not_ a goal of this project to generate PHP that would be checked in to a repo, worked on by developers, interface with existing PHP, or be called from PHP (although the latter two are _possible_, there are hoops you'd need to jump through since many of the data types are "boxed").

In the same way that the generated code coming out of Babel is readable but still kinda clunky, which is expected, that's kind of the expectation here.

So with that being said:

> The knowledge about this invariant is needed everywhere [...] As a consequence [...] it is very easy to break the implementation unintentionally
> Being forced to always use float literals also for counters [...] feels sort of cumbersome: `for ($i = 0.0; $i < 2.0; $i++) {}`

Yes, I agree with you @imme64, it's cumbersome, but as for the part about unintentionally breaking this invariant, I'd challenge that by asking: in what situation would you be writing code such that you'd need to remember or even think about this invariant? I realize I don't know much about your use case, but presumably you wouldn't be writing this PHP by hand, you'd be writing JS and generating the PHP, so you shouldn't need to "[be] forced to always use float literals", right?


> What about a flag that let's the user choose? In one mode everything is casted to float. Possibly with another flag to not have casts

If I understand you correctly @flip111, you're proposing two different compilation modes, on in which we do what we're doing now, and the other where it tries to leave integers as is but overrides equality checks (and probably other things) so that ints and floats can be used interchangeably in the PHP code.

This is probably doable, but would be a non-trivial amount of work and would require 2x test cases in many situations. I'd encourage you to weigh out the pros/cons before investing that effort, consider if it's just for more readable PHP or if it solves more problems in your use case.

In conclusion, to wrap up this issue, I'd say we do the simplest thing first, if you care about the 0xFF notation, and just modify the transpiler to output the `(float)0xFF` syntax, which would keep all the behavior the same as it is now.
